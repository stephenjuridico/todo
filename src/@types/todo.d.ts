import {RouteProp} from '@react-navigation/native';
import {Priority, SortBy, Status} from '../constants/constants';
import type {NativeStackNavigationProp} from '@react-navigation/native-stack';

export interface ITodo {
  id: string;
  title: string;
  description: string;
  status: Status;
  priority: Priority;
  createdAt?: string;
  updatedAt?: string;
}

export interface IFilters {
  status: string;
  name: SortBy;
  priority: SortBy;
}

export interface Response {
  message: string;
  item?: ITodo;
  error?: any;
}

export type TodoContextType = {
  data: ITodo[];
  mapped: MappedTodoType;
  filters: IFilters;
  isLoading: boolean;
  createTodo: (todo: ITodo) => Promise<Response>;
  updateTodo: (todo: ITodo) => Promise<Response>;
  updateStatus: (
    id: string,
    status: Status,
    showToast?: boolean,
  ) => Promise<Response>;
  setFilters: (filters: IFilters) => void;
};

export type MappedTodoType = {[key: string]: {count: number; items: ITodo[]}};

export type RootStackParamList = {
  Main: undefined;
  CreateTask: {item: ITodo} | undefined;
};

export type RootRouteProps<RouteName extends keyof RootStackParamList> =
  RouteProp<RootStackParamList, RouteName>;

export type MainScreenNavigationProp = NativeStackNavigationProp<
  RootStackParamList,
  'Main'
>;
