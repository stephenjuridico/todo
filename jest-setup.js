jest.mock('@react-native-async-storage/async-storage', () =>
  require('@react-native-async-storage/async-storage/jest/async-storage-mock'),
);

export const mockedNavigate = jest.fn();
export const mockedRoute = jest.fn();
export const mockedGoBack = jest.fn();
export const mockedRouteParam = jest.fn().mockReturnValue({
  params: {item: undefined},
});

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
      goBack: mockedGoBack,
    }),
    useRoute: () => mockedRouteParam,
  };
});

beforeEach(() => {
  jest.useFakeTimers();
});
