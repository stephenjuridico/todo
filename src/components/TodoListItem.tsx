import React, {FC} from 'react';
import {
  Box,
  HStack,
  Heading,
  Icon,
  Pressable,
  Stack,
  Text,
  VStack,
  Menu,
} from 'native-base';

import {PriorityColorCode, Status} from '../constants/constants';
import {useNavigation} from '@react-navigation/native';
import {ITodo, MainScreenNavigationProp} from '../@types/todo.d';

import Feather from 'react-native-vector-icons/Feather';
import {Alert} from 'react-native';
import {useTodo} from '../context/TodoContext';

interface TodoListItemProps {
  item: ITodo;
}

const TodoListItem: FC<TodoListItemProps> = ({item}) => {
  const {updateStatus} = useTodo();
  const navigation: MainScreenNavigationProp = useNavigation();
  const isCompleted = item.status === Status.Completed;
  const colorScheme = PriorityColorCode[item.priority];

  const trigger = (triggerProps: {_props: any; state: {open: boolean}}) => {
    return (
      <Pressable testID="btnTLIActions" {...triggerProps}>
        <Icon as={Feather} name="more-horizontal" />
      </Pressable>
    );
  };

  return (
    <Box bg="white" shadow={1} rounded="lg" mx={1} my={2} p={2}>
      <Stack space={2}>
        <HStack space={4} justifyContent={'space-between'}>
          <VStack flex={1}>
            <HStack space={2} alignItems={'center'}>
              <Heading
                strikeThrough={isCompleted}
                numberOfLines={1}
                size={'sm'}>
                {item.title}
              </Heading>
              <Icon as={Feather} name="flag" color={`${colorScheme}.500`} />
            </HStack>
            <Text color={'gray.400'} numberOfLines={2}>
              {item.description}
            </Text>
          </VStack>
          <Menu placement="left top" trigger={trigger}>
            <Menu.Item
              testID="btnTLIActionsEdit"
              onPress={() => navigation.navigate('CreateTask', {item})}>
              Edit
            </Menu.Item>
            <Menu.Item
              testID="btnTLIActionsUpdate"
              onPress={() =>
                updateStatus(
                  item.id,
                  item.status === Status.Pending
                    ? Status.Completed
                    : Status.Pending,
                  true,
                )
              }>
              {item.status === Status.Pending
                ? 'Mark as Done'
                : 'Mark as Pending'}
            </Menu.Item>
            {item.status !== Status.Archived && (
              <Menu.Item
                testID="btnTLIActionsArchive"
                onPress={() => {
                  Alert.alert(
                    'Archive Task',
                    'Are you sure you want to archive this task?',
                    [
                      {text: 'Cancel', style: 'cancel', onPress: () => {}},
                      {
                        text: 'Delete Task',
                        style: 'destructive',
                        onPress: () =>
                          updateStatus(item.id, Status.Archived, true),
                      },
                    ],
                  );
                }}>
                Archive
              </Menu.Item>
            )}
          </Menu>
        </HStack>
      </Stack>
    </Box>
  );
};

export default TodoListItem;
