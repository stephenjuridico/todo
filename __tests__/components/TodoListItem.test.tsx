import 'react-native';
import React from 'react';
import TodoListItem from '../../src/components/TodoListItem';
import {fireEvent, render, screen, waitFor} from '../../src/test/test-utils';
import {ITodo, TodoContextType} from '../../src/@types/todo';
import {
  TodoProvider,
  initialMappedValue,
  initialFiltersValue,
} from '../../src/context/TodoContext';
import {mockedNavigate} from '../../jest-setup';
import {Alert} from 'react-native';

describe('TodoListItem', () => {
  const expectedItem = {
    id: '123123',
    title: 'Test',
    description: 'Test Description',
    priority: 3,
    status: 'pending',
  } as ITodo;

  const mockedUpdateStatus = jest.fn();

  const renderTodo = () => {
    return render(
      <TodoProvider
        value={
          {
            data: [],
            mapped: initialMappedValue,
            isLoading: false,
            createTodo: jest.fn(),
            updateTodo: jest.fn(),
            updateStatus: mockedUpdateStatus,
            filters: initialFiltersValue,
            setFilters: jest.fn(),
          } as TodoContextType
        }>
        <TodoListItem item={expectedItem} />
      </TodoProvider>,
    );
  };

  it('renders correctly', async () => {
    renderTodo();
    await waitFor(() => {
      expect(screen.toJSON()).toMatchSnapshot();
    });
  });

  it('renders correctly and show more actions', async () => {
    renderTodo();
    const btnTLIActions = screen.getByTestId('btnTLIActions');
    fireEvent.press(btnTLIActions);
    const btnTLIActionsEdit = screen.getByTestId('btnTLIActionsEdit');
    const btnTLIActionsUpdate = screen.getByTestId('btnTLIActionsUpdate');
    const btnTLIActionsArchive = screen.getByTestId('btnTLIActionsArchive');
    expect(btnTLIActionsEdit).toBeTruthy();
    expect(btnTLIActionsUpdate).toBeTruthy();
    expect(btnTLIActionsArchive).toBeTruthy();
  });

  it('renders correctly and navigates to edit', async () => {
    renderTodo();
    const btnTLIActions = screen.getByTestId('btnTLIActions');
    fireEvent.press(btnTLIActions);
    const btnTLIActionsEdit = screen.getByTestId('btnTLIActionsEdit');
    fireEvent.press(btnTLIActionsEdit);
    expect(mockedNavigate).toBeCalled();
  });

  it('renders correctly and updates status', async () => {
    renderTodo();
    const btnTLIActions = screen.getByTestId('btnTLIActions');
    fireEvent.press(btnTLIActions);
    const btnTLIActionsUpdate = screen.getByTestId('btnTLIActionsUpdate');

    fireEvent.press(btnTLIActionsUpdate);

    expect(mockedUpdateStatus).toBeCalled();
  });

  it('renders correctly and archives item', async () => {
    // const errorMessageText = {
    //   title: 'Archive Task',
    //   message: 'Are you sure you want to archive this task?',
    // };
    // jest
    //   .spyOn(Alert, 'alert')
    //   .mockImplementation(('title', 'message', [{tex: 'Cancel', style: 'cancel', onPress: jest.fn()}]) => {});

    renderTodo();
    Alert.alert = jest.fn();
    const btnTLIActions = screen.getByTestId('btnTLIActions');
    fireEvent.press(btnTLIActions);
    const btnTLIActionsArchive = screen.getByTestId('btnTLIActionsArchive');
    fireEvent.press(btnTLIActionsArchive);
    // expect(
    //   screen.getByText('Are you sure you want to archive this task?'),
    // ).toBeTruthy();
    // expect(Alert.alert).toHaveBeenCalledWith(
    //   'Archive Task',
    //   'Are you sure you want to archive this task?',
    //   [
    //     {text: 'Cancel', style: 'cancel', onPress: jest.fn()},
    //     {
    //       text: 'Delete Task',
    //       style: 'destructive',
    //       onPress: jest.fn(),
    //     },
    //   ],
    // );

    // expect(mockedUpdateStatus).toBeCalled();
  });
});
