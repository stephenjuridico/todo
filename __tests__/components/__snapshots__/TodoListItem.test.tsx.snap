// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListItem renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    dataSet={{}}
    style={
      {
        "backgroundColor": "#FFFFFF",
        "borderRadius": 8,
        "elevation": 2,
        "marginBottom": 8,
        "marginLeft": 4,
        "marginRight": 4,
        "marginTop": 8,
        "paddingBottom": 8,
        "paddingLeft": 8,
        "paddingRight": 8,
        "paddingTop": 8,
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.2,
        "shadowRadius": 1.41,
      }
    }
  >
    <View
      dataSet={{}}
      style={{}}
    >
      <View
        dataSet={{}}
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <View
          dataSet={{}}
          style={
            {
              "flex": 1,
            }
          }
        >
          <View
            dataSet={{}}
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <Text
              dataSet={{}}
              numberOfLines={1}
              style={
                {
                  "backgroundColor": undefined,
                  "color": "#171717",
                  "fontFamily": undefined,
                  "fontSize": 16,
                  "fontStyle": "normal",
                  "fontWeight": "700",
                  "letterSpacing": 0,
                  "lineHeight": 20,
                  "textDecorationLine": undefined,
                }
              }
            >
              Test
            </Text>
            <View
              dataSet={{}}
              style={
                {
                  "width": 8,
                }
              }
            />
            <Text
              allowFontScaling={false}
              dataSet={{}}
              selectable={false}
              style={
                [
                  {
                    "color": undefined,
                    "fontSize": 12,
                  },
                  {
                    "color": "#ef4444",
                    "fontSize": 16,
                    "height": 16,
                    "lineHeight": 16,
                    "width": 16,
                  },
                  {
                    "fontFamily": "Feather",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
            >
              
            </Text>
          </View>
          <View
            dataSet={{}}
            style={{}}
          />
          <Text
            dataSet={{}}
            numberOfLines={2}
            style={
              {
                "backgroundColor": undefined,
                "color": "#a1a1aa",
                "fontFamily": undefined,
                "fontSize": 14,
                "fontStyle": "normal",
                "fontWeight": "400",
                "letterSpacing": 0,
                "lineHeight": 21,
                "textDecorationLine": undefined,
              }
            }
          >
            Test Description
          </Text>
        </View>
        <View
          dataSet={{}}
          style={
            {
              "width": 16,
            }
          }
        />
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          aria-haspopup="menu"
          collapsable={false}
          dataSet={{}}
          focusable={true}
          nativeID="1"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDownCapture={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={{}}
          testID="btnTLIActions"
        >
          <Text
            allowFontScaling={false}
            dataSet={{}}
            selectable={false}
            style={
              [
                {
                  "color": undefined,
                  "fontSize": 12,
                },
                {
                  "color": "#737373",
                  "fontSize": 16,
                  "height": 16,
                  "lineHeight": 16,
                  "width": 16,
                },
                {
                  "fontFamily": "Feather",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
